### PR TITLE
Add support functions for the new EVM intrinsics

### PR DESCRIPTION
**Pull request by elmato**

### CODE REVIEW

File: include/fc/crypto/blake2.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    enum class blake2b_error : int32_t {
+        none = 0,
+        input_len_error
+    };
+
+    std::pair<blake2b_error, bytes> blake2b(uint32_t _rounds, const bytes& _h, const bytes& _m, const bytes& _t0_offset, const bytes& _t1_offset, bool _f, const yield_function_t& yield);
+
+    typedef struct blake2b_state__ {
+        uint64_t h[8];
```
Comment (larryk85) [inline include/fc/crypto/blake2.hpp:22]: Adding default initialization here will help the compiler optimize those instructions and reschedule them best.

`uint64_t h[8] = {0,0,0,0,0,0,0,0};`
`uint64_t t[2]  = {0,0};`
`uint64_t f[2]  = {0,0};`

File: src/crypto/blake2.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    bytes out(sizeof(state.h), 0);
+
+    //  EIP-152 [4 bytes for rounds][64 bytes for h][128 bytes for m][8 bytes for t_0][8 bytes for t_1][1 byte for f] : 213
+    //          [------------------][64 bytes for h][128 bytes for m][8 bytes for t_0][8 bytes for t_1][------------] : 208
+    //  * rounds and final indicator flag are not serialized
+    if (_h.size() != 64 || _m.size() != 128 || _t0_offset.size() != 8 || _t1_offset.size() != 8) {
+        return std::make_pair(blake2b_error::input_len_error, out);
+    }
+
+    memset(&state, 0, sizeof(blake2b_state));
```
Comment (larryk85) [inline src/crypto/blake2.cpp:117]: This can be removed if you add initialization for `blake2b_state`.

File: include/fc/crypto/blake2.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        uint64_t t[2];
```
Comment (larryk85) [inline include/fc/crypto/blake2.hpp:23]: Is there a reason that `f` is 16 bytes wide?  It seems to only be setting 8 bytes.

File: src/crypto/blake2.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    v[8] = blake2b_IV[0];
+    v[9] = blake2b_IV[1];
+    v[10] = blake2b_IV[2];
+    v[11] = blake2b_IV[3];
+    v[12] = blake2b_IV[4] ^ S->t[0];
+    v[13] = blake2b_IV[5] ^ S->t[1];
+    v[14] = blake2b_IV[6] ^ S->f[0];
+    v[15] = blake2b_IV[7] ^ S->f[1];
```
Comment (larryk85) [inline src/crypto/blake2.cpp:96]: This seems to always be zero.

File: src/crypto/ecrecover.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    const secp256k1_context* ecrecover_context() {
+        static secp256k1_context* ctx = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
+        return ctx;
+    }
+
+    std::pair<ecrecover_error, bytes> ecrecover(const bytes& signature, const bytes& digest) {
+        const secp256k1_context* context{ecrecover_context()};
+        if (!context) {
+            return std::make_pair(ecrecover_error::init_error, bytes{});
```
Comment (larryk85) [inline src/crypto/ecrecover.cpp:16]: Seems like this should hard assert (throw exception) as I'm not sure how we can recover from that.

File: src/crypto/ecrecover.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        if (!secp256k1_ecdsa_recoverable_signature_parse_compact(context, &sig, (const unsigned char*)&signature[1], recid)) {
+            return std::make_pair(ecrecover_error::invalid_signature, bytes{});
+        }
+
+        secp256k1_pubkey pub_key;
+        if (!secp256k1_ecdsa_recover(context, &pub_key, &sig, (const unsigned char*)&digest[0])) {
+            return std::make_pair(ecrecover_error::recover_error, bytes{});
+        }
+
+        size_t kOutLen{65};
```
Comment (larryk85) [inline src/crypto/ecrecover.cpp:37]: Add `const` or `constexpr`.

File: src/crypto/modular_arithmetic.cpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    std::pair<modular_arithmetic_error, bytes> modexp(const bytes& _base, const bytes& _exponent, const bytes& _modulus)
+    {
+        if (_modulus.size() == 0) {
+            return std::make_pair(modular_arithmetic_error::modulus_len_zero, bytes{});
+        }
+
+        auto output = bytes(_modulus.size(), '\0');
+
+        mpz_t base;
+        mpz_init(base);
```
Comment (larryk85) [inline src/crypto/modular_arithmetic.cpp:16]: You can do something like:
```c++
mpz_t base, exponent, modulus;
mpz_inits(base, exponent, modulus, nullptr);
```

And then at the clear.
```c++
mpz_clears(base, exponent, modulus, nullptr);
```

Should be slightly faster as its grabbing memory at once and freeing the memory at once.

**Review by larryk85 — CHANGES_REQUESTED** [review]

**Review by larryk85 — CHANGES_REQUESTED** [review]
SHA3IUF submodule is still present.  I will approve after that one small issue is addressed.  The bigger concerns of how to manage GMP is beyond the scope of this and can be amended after this is merged.

**Review by larryk85 — DISMISSED** [review]

File: src/crypto/blake2.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+   - CC0 1.0 Universal : http://creativecommons.org/publicdomain/zero/1.0
+   - OpenSSL license   : https://www.openssl.org/source/license.html
+   - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
+
+   More information about the BLAKE2 hash function can be found at
+   https://blake2.net.
+*/
+
+#include <stdint.h>
+#include <string.h>
```
Comment (spoonincode) [inline src/crypto/blake2.cpp:17]: This file needs to include <limits>, it doesn't build on gcc12 for me otherwise

File: src/crypto/sha3.cpp [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+inline uint64_t rotl64(uint64_t x, uint64_t y) { return __builtin_rotateleft64(x, y); }
+#else
+// gcc should recognize this better than clang
+__attribute__ ((always_inline))
+inline uint64_t rotl64(uint64_t x, uint64_t y) { return (x << y) | (x >> (64-y)); }
+#endif
+/* This implementation is an amalgam from tiny_sha3 (https://github.com/mjosaarinen/tiny_sha3) and
+ * SHA3UIF (https://github.com/brainhub/SHA3UIF) and documentation.
+ * This implementation is quite slow comparative to openssl 1.1.1. Once we require a version greater
+ * than or equal to 1.1.1, we should replace with their primitives.
```
Comment (spoonincode) [inline src/crypto/sha3.cpp:39]: what's this comment about. mandel requires openssl 1.1+ now

File: src/crypto/blake2.cpp (hunk 8 above)
Comment (larryk85) [reply src/crypto/blake2.cpp:17]: You beat me to it. I just now tried to build it too.

Also, its small but change out <stdint.h> and <string.h> to <cstdint> and <cstring> accordingly to ensure that it is including from libcxx.

**Review by larryk85 — APPROVED** [review]

**Review by larryk85 — APPROVED** [review]

**Pull request merged**